### PR TITLE
NullPointerException fix

### PR DESCRIPTION
**Pull request by timzijl**
The NullPointerException in the server is fixed.

Server crashes with NullPointer #33

### CODE REVIEW

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 1]
```diff
@@ -128,85 +128,89 @@ public void run() {
...
-				
+
 				Log.DEBUG.println("Closing client socket connection");
 				client.close();
 			}
+
+			key.cancel();
+			key.attach(null);
 		} catch (IOException e) {
 			;
```
Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Exception catching without doing anything? at least write something to the log

Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Same thing: Exception catching without doing anything? at least write something to the log

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 2]
```diff
@@ -243,37 +247,37 @@ public void exit() {
 		} catch (Exception e) {
 			;
```
Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Same

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 3]
```diff
@@ -243,37 +247,37 @@ public void exit() {
...
 		}
-		
+
 		try {
 			Log.DEBUG.println("Closing client input handler");
 			clientInputHandler.exit();
 		} catch (Exception e) {
 			;
```
Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Same

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 4]
```diff
@@ -243,37 +247,37 @@ public void exit() {
...
 		}
-		
+
 		try {
 			Log.DEBUG.println("Closing server selector");
 			selector.close();
 		} catch (IOException e) {
 			;
```
Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Same

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 5]
```diff
@@ -243,37 +247,37 @@ public void exit() {
...
 		}
-		
+
 		try {
 			Log.DEBUG.println("Closing server socket");
 			serverSocketChannel.close();
 		} catch (IOException e) {
 			;
```
Comment (IcyPalm) [inline src/nl/hanze/gameserver/server/GameServer.java]: Same

**Comment by IcyPalm** [comment]
Please do something with the empty catches, that is just very fugly(VERY FUGLY), I know it is the code as was delivered by the Hanze but let's show that we can do it better :+1:

**Comment by timzijl** [comment]
Dear Mr Pool,

Those are all shit from the Hanze university of applied sciences. The things I added to a try I also catched and printed why it was happening.

Sincerely,

Tim Zijlstra

**Comment by roffel** [comment]
Just print the error text and it will be fine I think. Sorry Tim <3

**Comment by IcyPalm** [comment]
> Just print the error text and I think it will be fine
Exactly this is what I meant

**Comment by roffel** [comment]
We still love you Tim.

File: src/nl/hanze/gameserver/server/ClientManager.java [hunk 6]
```diff
@@ -147,59 +150,61 @@ public void message(Client player, Client opponent, String chatText) {
...
 	}
-	
+
 	public void cancelChallenge(Challenge challenge) {
 		cancelChallenge(challenge, true);
 	}
-	
+
 	private void cancelChallenge(Challenge challenge, boolean notify) {
 		if(challenge == null) {
```
Comment (Jarinus) [inline src/nl/hanze/gameserver/server/ClientManager.java]: Replace this part with `if (challenge != null)` instead of the `challenge == null` check, then nothing inside the if-statement and using the else keyword...

File: src/nl/hanze/gameserver/server/ClientManager.java [hunk 7]
```diff
@@ -147,59 +150,61 @@ public void message(Client player, Client opponent, String chatText) {
...
-		
-		challenge.getPlayer().setChallenge(null);
-		challenge.getOpponent().removeChallenge(challenge);
-		
-		if(notify) {
-			String challengeNumberString = StringUtils.toStringAsMap("CHALLENGENUMBER", challenge.getChallengeNumber());
-			String format = "CHALLENGE CANCELLED %s";
-			challenge.getOpponent().writeResponse(new GameResponse(String.format(format, challengeNumberString)));
+		else {
+
```
Comment (Jarinus) [inline src/nl/hanze/gameserver/server/ClientManager.java]: Please refrain from adding unnecessary empty lines. You've also got loads of unnecessary changes. Please try to minimize that.

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 8]
```diff
@@ -217,19 +221,19 @@ public void disconnect(SocketChannel client) {
...
 	}
-	
+
 	public void exit() {
 		running = false;
-		
+
 		try {
 			for(SelectionKey key: selector.keys()) {
 				try {
```
Comment (Jarinus) [inline src/nl/hanze/gameserver/server/GameServer.java]: Why would you put a try-catch block within a try-catch block..?

File: src/nl/hanze/gameserver/server/GameServer.java [hunk 9]
```diff
@@ -217,19 +221,19 @@ public void disconnect(SocketChannel client) {
...
-		
+
 		synchronized(opsChangeLock) {
 			selector.wakeup();
 			key.interestOps(key.interestOps() | SelectionKey.OP_WRITE);
 		}
 	}
-	
+
 	public void exit() {
```
Comment (Jarinus) [inline src/nl/hanze/gameserver/server/GameServer.java]: It seems that you've got a LOT of try-catch blocks in this method. That's not a nice way to do it.

**Comment by Jarinus** [comment]
The code should work. Whether it's neat code shouldn't be the reason to merge it or not, so I'll give it a GO.

**Comment by timzijl** [comment]
Dear Jari, do it yourself or shut up. This is the shitty code from the Hanze

**Comment by Jarinus** [comment]
Wow, offensive much?

**Comment by goto-bus-stop** [comment]
Thanks for digging into this! :D :tada: 

(Discussed privately w/ @IcyPalm & LGTM'd)

> Dear Jari, do it yourself or shut up.

The goal of a pull request is to improve the code (through reviews by other people) and _then_ merge it. Don't take comments on your code as personal attacks--they're there to help make everyone's lives better :sparkles: 

---

@Jarinus A bunch of your comments were on code that @timzijl didn't touch--i.e., code that is not relevant to this PR. If you disagree with how unrelated parts of the code work, that's totally cool, but perhaps better suited for a separate issue or PR :)

---

(General comment re: empty line changes: clearing all of the whitespace-only lines makes for really noisy diffs that are hard to review, but it's what most editors do for autoformatting by default (and righteously so), and thus hard to avoid. Really, we should've run that autoformat on the entire codebase at the start, but it's a bit late for that now :smile: )
